### PR TITLE
#114 Add in memory preference

### LINKED ISSUE #114: Add MemoryPreference artifact?
I have rewritten MemoryPreference a few times in various projects for tests and mock mode.

```java
public final class MemoryPreference<T> implements Preference<T> {
  private final String key;
  private final T defaultValue;
  @Nullable private volatile T value;
  ...
}
```

Maybe there should be an artifact here for it?

### PR DESCRIPTION
**Pull request by kirillzh**
Adding in-memory implementation of the `Preference` that can be used for testing purposes.
Closes #114 

Here's a primitive example of how and why this can be used. Let's say we have some dummy class that is backed by a boolean flag and boolean `Preference`:
```kotlin
class Foo(val pref: Preference<Boolean>) {
    private var localFlag: Boolean = false

    val isEnabled: Boolean get() = localFlag && pref.get()

    fun enable() = set(true)
    fun disable() = set(false)

    private fun set(state: Boolean) {
        this.localFlag = state
        pref.set(state)
    }
}
```

To test `Foo#isEnabled` property we need some implementation of the `Preference` which would persist and change its state accordingly:

```kotlin
val pref = TestPreference("key", false)
val foo = Foo(pref)

assertThat(foo.isEnabled).isFalse()
foo.enable()
assertThat(foo.isEnabled).isTrue()
foo.disable()
assertThat(foo.isEnabled).isFalse()
```

Some remaining tasks:
- [ ] add usage documentation
- [ ] convert back to Java, maybe?

### CODE REVIEW

**Comment by kirillzh** [comment]
Once #124 is merged to master by the owners, unnecessary commits should go away from this PR.

**Comment by kirillzh** [comment]
Any suggestions where `TestPreference.kt` should be located? `:rx-preferences:androidJavadocs` didn't like it when I put the file into `test` package.

File: rx-preferences/src/main/java/com/f2prateek/rx/preferences2/test/TestPreference.kt [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import io.reactivex.functions.Consumer
+import io.reactivex.subjects.BehaviorSubject
+
+/**
+ * In-memory implementation of [Preference] that can be used for testing purposes. Only use this
+ * if your code under test depends on the changes in the preferences. Otherwise, if you need to
+ * pass the preference as a dependency and don't care about its state, simply mock the preference
+ * using [Preference] interface.
+ */
+class TestPreference<T>(private val key: String, private val defaultValue: T) : Preference<T> {
```
Comment (kirillzh) [inline rx-preferences/src/main/java/com/f2prateek/rx/preferences2/test/TestPreference.kt]: Any suggestions for name?

File: rx-preferences/src/main/java/com/f2prateek/rx/preferences2/test/TestPreference.kt [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import com.f2prateek.rx.preferences2.Preference
+import io.reactivex.Observable
+import io.reactivex.functions.Consumer
+import io.reactivex.subjects.BehaviorSubject
+
+/**
+ * In-memory implementation of [Preference] that can be used for testing purposes. Only use this
+ * if your code under test depends on the changes in the preferences. Otherwise, if you need to
+ * pass the preference as a dependency and don't care about its state, simply mock the preference
+ * using [Preference] interface.
```
Comment (kirillzh) [inline rx-preferences/src/main/java/com/f2prateek/rx/preferences2/test/TestPreference.kt]: Would love to get any corrections, if any!

File: rx-preferences/src/main/java/com/f2prateek/rx/preferences2/test/TestPreference.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    override fun isSet(): Boolean = value != null
+
+    override fun delete() {
+        this.value = null
+        values.onNext(defaultValue)
+    }
+
+    override fun asObservable(): Observable<T> = values.distinctUntilChanged()
```
Comment (kirillzh) [inline rx-preferences/src/main/java/com/f2prateek/rx/preferences2/test/TestPreference.kt]: Was thinking if I should store `values.distinctUntilChanged()` into a property and store it here.

File: rx-preferences/src/main/java/com/f2prateek/rx/preferences2/test/TestPreference.kt [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+     * @see [isSet].
+     * @see [get].
+     */
+    private var value: T? = null
+    /**
+     * Allows observing changes on this preference, starts with [defaultValue].
+     *
+     * @see [asObservable].
+     */
+    private val values = BehaviorSubject.createDefault<T>(defaultValue)
```
Comment (kirillzh) [inline rx-preferences/src/main/java/com/f2prateek/rx/preferences2/test/TestPreference.kt]: Any RxJava edge cases that I should look out for?

File: rx-preferences/src/main/java/com/f2prateek/rx/preferences2/RxSharedPreferences.java [hunk 5]
```diff
@@ -19,6 +20,7 @@
 import static com.f2prateek.rx.preferences2.Preconditions.checkNotNull;
 
 /** A factory for reactive {@link Preference} objects. */
+@SuppressWarnings("WeakerAccess")
```
Comment (kirillzh) [inline rx-preferences/src/main/java/com/f2prateek/rx/preferences2/RxSharedPreferences.java]: Removes warnings about public APIs not being used, I wish Java had something similar to Kotlin's `@PublishedApi`...

File: rx-preferences/build.gradle [hunk 6]
```diff
@@ -20,12 +21,15 @@ android {
 }
 
 dependencies {
-  compile deps.annotations
-  compile deps.rxjava
+  implementation deps.annotations
+  implementation deps.kotlinStdlib
```
Comment (kirillzh) [inline rx-preferences/build.gradle]: Should this be public? `TestPreference.kt` doesn't expose any APIs from stdlib.

**Comment by kirillzh** [comment]
> Any suggestions where `TestPreference.kt` should be located? `:rx-preferences:androidJavadocs` didn't like it when I put the file into `test` package.

It's not package name, it's file name (or both)? Any suggestions? Not sure how to tune it in to allow `test/TestPreference.kt` naming...

**Comment by kirillzh** [comment]
Considering that I created only one new file, adding Kotlin as a new dependency probably wasn't a good idea, can revert to Java. Thoughts?

File: rx-preferences/src/main/java/com/f2prateek/rx/preferences2/TestPreference.kt [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+import io.reactivex.functions.Consumer
+import io.reactivex.subjects.BehaviorSubject
+
+/**
+ * In-memory implementation of [Preference] that can be used for testing purposes. Only use this
+ * if your code under test depends on the changes in the preferences. Otherwise, if you need to
+ * pass the preference as a dependency and don't care about its state, simply mock the preference
+ * using [Preference] interface.
+ */
+class TestPreference<T>(private val key: String, private val defaultValue: T) : Preference<T> {
```
Comment (f2prateek) [inline rx-preferences/src/main/java/com/f2prateek/rx/preferences2/TestPreference.kt]: How about naming this `MemoryPreference`?

File: build.gradle [hunk 8]
```diff
@@ -1,50 +1,59 @@
...
-    mavenCentral()
-    jcenter()
-  }
+    ext {
+        buildToolsVersion = '28.0.3'
+        butterknifeVersion = '10.1.0'
+        compileSdkVersion = 28
+        java8Version = JavaVersion.VERSION_1_8
+        javaVersion = JavaVersion.VERSION_1_7
+        kotlinVersion = '1.3.21'
```
Comment (f2prateek) [inline build.gradle]: I'm not sure we want to add a Kotlin dependency just for this.
